flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunDevBundleBuild.java

```diff
@@ -82,10 +89,155 @@ public static boolean needsBuild(File npmFolder) {
+        }
+
+        JsonObject bundleModules = statsJson.getObject("npmModules");
+
+        // Check that bundle modules contains all package dependencies
+        if (packageJsonHash.equals(bundlePackageJsonHash)) {
+            if (missingNpmPackages(npmPackages, bundleModules)) {
+                return false;
+            }
+        }
+
+        JsonObject dependencies = packageJson.getObject("dependencies");
+
+        List<String> collection = Arrays.stream(dependencies.keys())
+                .filter(pkg -> !bundleModules.hasKey(pkg))
+                .collect(Collectors.toList());
+
+        if (!collection.isEmpty()) {
+            for (String dependency : collection) {
+                getLogger().info("Dependency " + dependency
+                        + " is missing from the bundle");
+            }
+            return false;
+        }
+
+        // We know here that all dependencies exist
+        collection = Arrays.stream(dependencies.keys())
+                .filter(pkg -> !dependencies.getString(pkg)
+                        .equals(bundleModules.getString(pkg)))
+                .collect(Collectors.toList());
+
+        if (!collection.isEmpty()) {
+            for (String pkg : collection) {
+                getLogger().info(
+                        "Dependency {}:{} has the wrong version {} in the bundle",
+                        pkg, dependencies.getString(pkg),
+                        bundleModules.getString(pkg));
+            }
+            return false;
+        }
+
+        return true;
+    }
+
+    private static JsonObject getPackageJson(File npmFolder) {
+        File packageJsonFile = new File(npmFolder, "package.json");
+
+        if (packageJsonFile.exists()) {
+            try {
+                return Json.parse(FileUtils.readFileToString(packageJsonFile,
+                        StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                getLogger().warn("Failed to read package.json", e);
+            }
+        }
+        return null;
+    }
+
+    private static String getStatsHash(JsonObject statsJson) {
+        if (statsJson.hasKey("packageJsonHash")) {
+            return statsJson.getString("packageJsonHash");
+        }
+
+        return null;
+    }
+
+    private static String getPackageJsonHash(JsonObject packageJson) {
+        if (packageJson != null && packageJson.hasKey("vaadin")
+                && packageJson.getObject("vaadin").hasKey("hash")) {
+            return packageJson.getObject("vaadin").getString("hash");
+        }
+
+        return null;
+    }
+
+    /**
+     * Check that all npmPackages are in the given dependency object.
+     *
+     * @param npmPackages
+     *            {@code @NpmPackage} key-value map
+     * @param dependencies
+     *            json object containing dependencies to check against
+     * @return {@code false} if all packages are found
+     */
+    private static boolean missingNpmPackages(Map<String, String> npmPackages,
```

Comment:
```suggestion
    private static boolean dependenciesContainsAllPackages(Map<String, String> npmPackages,
```
+ inverse logic to be more readable

Reply:
done,